0.data_analysis_and_processing/nbconverted/filter_nuclei.py

```diff
@@ -0,0 +1,183 @@
+
+# In[1]:
+
+
+import pathlib
+import warnings
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import seaborn as sns
+
+
+# ## Find the root of the git repo on the host system
+
+# In[2]:
+
+
+# Get the current working directory
+cwd = pathlib.Path.cwd()
+
+if (cwd / ".git").is_dir():
+    root_dir = cwd
+
+else:
+    root_dir = None
+    for parent in cwd.parents:
+        if (parent / ".git").is_dir():
+            root_dir = parent
+            break
+
+# Check if a Git root directory was found
+if root_dir is None:
+    raise FileNotFoundError("No Git root directory found.")
+
+# Remove categorical warning in seaborn
+warnings.filterwarnings("ignore", category=FutureWarning)
+
+
+# # Inputs
+
+# In[3]:
+
+
+data_dir = root_dir / "nuclear_speckles_data"
+nuclear_mask_dir = (data_dir / "Nuclear_masks").resolve(strict=True)
+sc_profiles_path = list((data_dir / "Preprocessed_data/single_cell_profiles").resolve(strict=True).glob("*annotated*.parquet"))
+
+# Load single-cell profile data
+scdfs = [pd.read_parquet(sc_path) for sc_path in sc_profiles_path if sc_path.is_file()]
+
+
+# # Outputs
+
+# In[4]:
+
+
+filtered_sc_path = pathlib.Path("filtered_single_cells")
+filtered_sc_figure_path = filtered_sc_path / "filtered_sc_figures"
+
+filtered_sc_figure_path.mkdir(parents=True, exist_ok=True)
+
+
+# In[5]:
+
+
+def filter_bounding_box_size(_scdf, _bounding_box_col):
+    """
+    Filter nuclei if the robust z score of a bounding box dimension is less than three.
+
+    Parameters
+    ----------
+    _scdf: Pandas Dataframe
+        The unfiltered single-cell data containing bounding box dimensions.
+
+    _bounding_box_col: String
+        Column specifying the bounding box dimension.
+
+    Returns
+    -------
+    Filtered single-cell pandas dataframe.
+    """
+
+    median = scdfs[_bounding_box_col].median()
+
+    # Calculate the absolute deviations from the median
+    _scdf[f"{_bounding_box_col}_absolute_deviation"] = np.abs(_scdf[_bounding_box_col] - median)
+
+    # Calculate the MAD
+    mad = _scdf[f"{_bounding_box_col}_absolute_deviation"].median()
+
+    # Calculate the robust Z-score
+    _scdf[f"{_bounding_box_col}_robust_z_score"] = (_scdf[_bounding_box_col] - median) / mad
+
+    # Returns the single cells with a robust z score less than three
+    return _scdf.loc[_scdf[f"{_bounding_box_col}_robust_z_score"] < 3]
+
+
+# ## Retain only Common Metadata and Bounding Box Columns
+
+# In[6]:
+
+
+bounding_box_cols = ["Nuclei_AreaShape_BoundingBoxMaximum_X", "Nuclei_AreaShape_BoundingBoxMinimum_X", "Nuclei_AreaShape_BoundingBoxMaximum_Y", "Nuclei_AreaShape_BoundingBoxMinimum_Y"]
+
+common_cols = set(scdfs[0].columns[scdfs[0].columns.str.contains('Metadata')])
+
+for scdf in scdfs[1:]:
+    common_cols = set(scdf.columns[scdf.columns.str.contains('Metadata')]) & set(scdf.columns)
+
+scdfs = pd.concat(scdfs, axis=0)[list(common_cols) + bounding_box_cols]
```

Comment:
Is there a reason you have to specify common Metadata? From my understanding, the metadata columns are already consistent across the plates in this project. I would recommend that if this is not the case, add a code comment here to say which Metadata is consistent.

Reply:
I was thinking of exploring each plate's Metadata in a script, although, I decided it would be easier to just do this in the code. I also like this approach because if for some reason the data changes, or there is a bug, a reviewer of the code can have confidence the columns are the same by virtue of the code.